### PR TITLE
add CompoundFile::File std::istream ctor

### PR DESCRIPTION
**Pull request by ggeenn**

### CODE REVIEW

**Comment by igormironchik** [comment]
Hello. Thanks for PR. Could you answer why do you need ctor of `ConmpoundFile::File` with `std::istream`?

You added `old.xls` test Excel file. What BIFF version of this file?

**Comment by ggeenn** [comment]
Hello. The main issue is document in zip archive. It provides std::istream. So, using ConmpoundFile::File I have to save document into temp file before parsing.

I don't know about BIFF enough, but just can see that old.xls has primary stream "Book" instead of "Workbook" hardcoded in Book::loadBook.

**Comment by igormironchik** [comment]
> The main issue is document in zip archive. It provides std::istream. So, using ConmpoundFile::File I have to save document into temp file before parsing.

Ok, I see. Could you provide then corresponding API for `Excel::Book` with `std::istream`. `CompoundFile` is the only storage of streams. I never thought that somebody will use `CompoundFile` standalone.

> I don't know about BIFF enough, but just can see that old.xls has primary stream "Book" instead of "Workbook" hardcoded in Book::loadBook.

Please inspect your `old.xls` file for BIFF version. In `doc` folder you can find all necessary documentation about compound file and Excel. Have a look at `Excel97-2007BinaryFileFormat(xls)Specification.pdf` and search there for `BOF` record, there you can understand what BIFF version of your file. If BIFF less than 8, then, please, remove your test Excel file and test with it. This library supports only BIFF8 (Excel 97-2003), so hardcoded `Workbook` directory is from specification of BIFF8 format, as I remember, I wrote this library very long time ago, so already forgot all niceties of the Excel file format, please, correct me if I'm wrong. Thanks.

**Comment by igormironchik** [comment]
This is very strange Excel file. It reports that use BIFF8, but doesn't have `Workbook` directory, but `Book`. Where this file was created? I'm not sure that this file meets the BIFF8 standard. But LibreOffice and Google Sheets able to open this file...

I re-saved it in LibreOffice with Excel 97-2003 format, and my library then able to parse it.

**Comment by igormironchik** [comment]
Even Microsoft Office opens this file with some mess in content:

```
???? / ???? ?? ???? ?- zip ?? ?????? ?????? ?? ???? ?- Somefile Mac ?????, SomefileMacAgentInstaller.zip ?????? ?- Macintosh ???.
??? ?? ???? ?- zip ??? ????? / ????? ?? ?????? ?? ?????? ???? Mac ?? Somefile, SomefileMacAgentInstaller.
??? ?? ???? ??????.
```
I guess that this file is a mess of different formats.

**Comment by igormironchik** [comment]
Yes, `old.xls` reports BIFF8, but uses LABEL record to store string, and this is pre-BIFF8. In BIFF8 for strings LABELSST uses.

**Comment by ggeenn** [comment]
> Ok, I see. Could you provide then corresponding API for `Excel::Book` with `std::istream`. `CompoundFile` is the only storage of streams. I never thought that somebody will use `CompoundFile` standalone.

The deal.

> Please inspect your `old.xls` file for BIFF version. In `doc` folder you can find all necessary documentation about compound file and Excel. Have a look at `Excel97-2007BinaryFileFormat(xls)Specification.pdf` and search there for `BOF` record, there you can understand what BIFF version of your file. If BIFF less than 8, then, please, remove your test Excel file and test with it. This library supports only BIFF8 (Excel 97-2007), so hardcoded `Workbook` directory is from specification of BIFF8 format, as I remember, I wrote this library very long time ago, so already forgot all niceties of the Excel file format, please, correct me if I'm wrong. Thanks.

version is BIFF8( = 0x0600). Should I just add alternative name (will have both 'Workbook' & 'Book')?

**Comment by igormironchik** [comment]
> version is BIFF8( = 0x0600). Should I just add alternative name (will have both 'Workbook' & 'Book')?

As I already said, your file is a mess of different standards. Please remove this file from PR and test using it.

File: read-excel/book.hpp [hunk 1]
```diff
@@ -175,13 +187,15 @@ Book::sheet( size_t index ) const
 }
 
 inline void
-Book::loadBook( const std::string & fileName )
+Book::loadBook( std::istream & fileStream, const std::string& fileName )
```
Comment (igormironchik) [inline read-excel/book.hpp]:
```suggestion
Book::loadBook( std::istream & fileStream, const std::string & fileName )
```

File: read-excel/book.hpp [hunk 2]
```diff
@@ -175,13 +187,15 @@ Book::sheet( size_t index ) const
...
 inline void
-Book::loadBook( const std::string & fileName )
+Book::loadBook( std::istream & fileStream, const std::string& fileName )
 {
 	try {
 		clear();
 
-		CompoundFile::File file( fileName );
-		auto stream = file.stream( file.directory( L"Workbook" ) );
+		CompoundFile::File file( fileStream, fileName);
```
Comment (igormironchik) [inline read-excel/book.hpp]:
```suggestion
		CompoundFile::File file( fileStream, fileName );
```

File: read-excel/book.hpp [hunk 3]
```diff
@@ -175,13 +187,15 @@ Book::sheet( size_t index ) const
...
+		auto stream = file.stream( 
+			file.hasDirectory( L"Workbook" ) ? file.directory( L"Workbook" )
```
Comment (igormironchik) [inline read-excel/book.hpp:197]: Point me to BIFF8 standard where `Book` is standardized in BIFF8, or remove this selection. This library supports BIFF8 only.

File: read-excel/compoundfile/compoundfile.hpp [hunk 4]
```diff
@@ -59,18 +59,28 @@ namespace CompoundFile {
...
 public:
+	explicit File( std::istream & stream, const std::string & fileName = "<custom-stream>" );
 	explicit File( const std::string & fileName );
 	~File();
 
 	//! \return Directory entry by its name.
 	Directory directory( const std::wstring & name ) const;
 
+	//! \return is Directory entry exist by its name.
+	bool hasDirectory(const std::wstring & name ) const;
```
Comment (igormironchik) [inline read-excel/compoundfile/compoundfile.hpp]:
```suggestion
	bool hasDirectory( const std::wstring & name ) const;
```

File: read-excel/compoundfile/compoundfile.hpp [hunk 5]
```diff
@@ -158,11 +168,64 @@ loadChildDirectories( std::vector< Directory > & dirs,
 // File
 //
 
+inline
+File::File( std::istream & stream, const std::string & fileName )
+	: m_stream( stream )
```
Comment (igormironchik) [inline read-excel/compoundfile/compoundfile.hpp]:
```suggestion
	:	m_stream( stream )
```

File: read-excel/compoundfile/compoundfile.hpp [hunk 6]
```diff
@@ -158,11 +168,64 @@ loadChildDirectories( std::vector< Directory > & dirs,
...
+File::File( std::istream & stream, const std::string & fileName )
+	: m_stream( stream )
+{
+	initialize( fileName );
+}
+
 inline
 File::File( const std::string & fileName )
+	: m_fileStream( fileName, std::ios::in | std::ios::binary )
+	, m_stream( m_fileStream )
```
Comment (igormironchik) [inline read-excel/compoundfile/compoundfile.hpp]:
```suggestion
	:	m_fileStream( fileName, std::ios::in | std::ios::binary )
	,	m_stream( m_fileStream )
```

File: test/CMakeLists.txt [hunk 7]
```diff
@@ -13,6 +13,8 @@ file( COPY data/test.xls
 	DESTINATION ${CMAKE_CURRENT_BINARY_DIR}/data )
 file( COPY data/verybig.xls
 	DESTINATION ${CMAKE_CURRENT_BINARY_DIR}/data )
+file( COPY data/old.xls
```
Comment (igormironchik) [inline test/CMakeLists.txt]: Remove `old.xls` from this PR, as this file doesn't feet a BIFF8 standard.

File: test/book/main.cpp [hunk 8]
```diff
@@ -66,3 +66,11 @@ TEST_CASE( "test_book" )
 	REQUIRE( std::fabs( sheet->cell( 2, 3 ).getFormula().getDouble() - 3.3 ) < 1E-9 );
 
 }
+
+TEST_CASE( "test_book_via_stream" )
+{
+	std::ifstream fileStream( "./test/data/old.xls", std::ios::in | std::ios::binary );
```
Comment (igormironchik) [inline test/book/main.cpp]: Use for test one of the existing files.

File: test/compoundfile/main.cpp [hunk 9]
```diff
@@ -144,3 +144,21 @@ TEST_CASE( "test_stream" )
...
 }
+
+
+//
+// test_stream
+//
+
+TEST_CASE( "test_stream_name" )
+{
+	std::ifstream fileStream( "./test/data/old.xls", std::ios::in | std::ios::binary );
```
Comment (igormironchik) [inline test/compoundfile/main.cpp]: Use for test one of the existing files.

**Review by igormironchik — CHANGES_REQUESTED** [review]

**Comment by igormironchik** [comment]
> A LABEL record describes a cell that contains a pre-BIFF8 string constant. Note:
this was replaced in BIFF8 by LABELSST.

But `old.xls` uses `LABEL` for string. I can't approve this test data file.

File: read-excel/book.hpp (hunk 3 above)
Comment (ggeenn) [reply read-excel/book.hpp:197]: ok, will be removed. ( But catdoc does the same here - https://github.com/petewarden/catdoc/blob/master/src/xls2csv.c#L166, so maybe this document isn't the one )

Comment (ggeenn) [reply read-excel/book.hpp:197]: Done

Comment (igormironchik) [reply read-excel/book.hpp:197]: Ok, I will approve checking for `Workbook` or `Book` to load.

**Review by igormironchik — APPROVED** [review]

**Pull request merged**